doc/developer-guides/sw_design_guidelines.rst

```diff
@@ -0,0 +1,494 @@
+-------------------------------------------
+
+The rules of error detection and error handling on module level is shown in
+:numref:`rules_module_level` below.
+
+.. table:: Rules of Error Detection and Error Handling on Module Level
+   :align: center
+   :widths: auto
+   :name: rules_module_level
+
+   +--------------------+-----------+----------------------------+---------------------------+-------------------------+
+   | **Resource Class** | **Failure | **Error Detection via      | **Error Handling Policy** | **Example**             |
+   |                    | Mode**    | Hypervisor**               |                           |                         |
+   +====================+===========+============================+===========================+=========================+
+   | Internal data of   | N/A       | Partial no.                | The hypervisor shall use  | virtual PCI device      |
+   | the hypervisor     |           | The related pre-conditions | the internal resource/data| information, defined    |
+   |                    |           | are required.              | directly.                 | with array 'pci_vdevs[]'|
+   |                    |           | The design will guarantee  |                           | through static          |
+   |                    |           | the correctness and the    |                           | allocation.             |
+   |                    |           | test cases will verify the |                           |                         |
+   |                    |           | related pre-conditions.    |                           |                         |
+   |                    |           | If the design can not      |                           |                         |
+   |                    |           | guarantee the correctness, |                           |                         |
+   |                    |           | the related error handling |                           |                         |
+   |                    |           | codes need to be added.    |                           |                         |
+   |                    |           | Note: Some examples of     |                           |                         |
+   |                    |           | pre-conditions are listed, |                           |                         |
+   |                    |           | like non-empty array, valid|                           |                         |
+   |                    |           | array size and non-null    |                           |                         |
+   |                    |           | pointer.                   |                           |                         |
+   +--------------------+-----------+----------------------------+---------------------------+-------------------------+
+   | Configuration data | Corrupted | No.                        | The bootloader initializes| 'vm_config->pci_ptdevs' |
```

Comment:
Tables like this are hard to edit, but this should read ``(including code, data, and bss)``, and in the last column the final period needs to adjacent to the last word, so it says ``statically.``